### ISSUE TITLE
[Bug] DeepSeek chunked-prefix prefill merges base-2 LSE into a natural-log merge_state_v2 (silent accuracy loss on every prefix-cache hit)

### ISSUE THREAD

**Issue by acmore** [sgl-project/sglang#40903]
### Summary

On the DeepSeek MLA chunked-prefix prefill path, the per-chunk log-sum-exp values produced by the ragged attention kernels are in **base 2**, while `merge_state_v2` (and the Triton fallback) combine them with a **natural-log** formula. Nothing converts in between, so every prefix-cache hit merges the chunks with the wrong weights: the long cached prefix is over-weighted by 2^(0.4427·Δ) relative to the new tokens.

This is silent — no error, no warning, plausible-looking output — and on Blackwell it is on by default, because `trtllm_mla` is the auto-selected backend for `DeepseekV3ForCausalLM` on sm100 and it takes the chunk-and-merge path on *every* extend with a non-empty prefix.

### Where

- `python/sglang/srt/models/deepseek_common/attention_forward_methods/forward_mha.py`, `_chunked_prefix_attn_mha()`: calls `merge_state(output, lse, accum_output, accum_lse, ...)` with the LSE exactly as the backend returned it.
- `python/sglang/srt/layers/attention/merge_state.py` → `sgl_kernel.merge_state_v2`, else `merge_state_triton`. The Triton kernel is explicit about the convention it assumes: `out_se = tl.exp(p_lse) + tl.exp(s_lse)`, `out_lse = tl.log(out_se) + max_lse`.
- No backend converts on the way out: `grep -n "log2\|1.4427\|LOG2E" python/sglang/srt/layers/attention/trtllm_mla_backend.py` is empty.

Verified on `main` at `2d25767`, and reproduced at runtime on 0.5.15.post1 and 0.5.19.

vLLM hits the same kernels and converts per backend, which is presumably why it does not show the problem: `log2_lse_to_ln` (defined in `vllm/v1/attention/backends/utils.py`) is applied to the returned LSE inside `vllm/v1/attention/backends/mla/prefill/flashinfer.py` and `.../trtllm_ragged.py`.

### Reproduction

One GPU, no model weights (DeepSeek-V3 MHA prefill shapes on a TP8 slice: 16 heads, qk head_dim 192, v head_dim 128, bf16):

```python
"""Standalone repro: the DeepSeek chunked-prefix path merges base-2 LSE with a
natural-log merge. One GPU, no model weights.

Part 1 - what base does each ragged-prefill kernel return its LSE in?
Part 2 - what does that do to the merged output, on a long-prefix / short-suffix split?

python repro_lse_base.py
"""

import math

import torch
import flashinfer
from sgl_kernel import merge_state_v2

torch.manual_seed(0)
H, DQK, DV = 16, 192, 128          # DeepSeek-V3 MHA prefill shapes on a TP8 slice
SCALE = 1.0 / math.sqrt(DQK)
LN2 = math.log(2)
ws = torch.zeros(256 * 1024 * 1024, dtype=torch.uint8, device="cuda")


def trtllm_ragged(q, k, v, causal=False):
    n, m = k.shape[0], q.shape[0]
    out = torch.empty(m, H, DV, dtype=torch.bfloat16, device="cuda")
    return flashinfer.prefill.trtllm_ragged_attention_deepseek(
        query=q, key=k, value=v, workspace_buffer=ws, batch_size=1, window_left=-1,
        enable_pdl=False, max_q_len=m, bmm1_scale=SCALE, bmm2_scale=1.0,
        cum_seq_lens_q=torch.tensor([0, m], dtype=torch.int32, device="cuda"),
        cum_seq_lens_kv=torch.tensor([0, n], dtype=torch.int32, device="cuda"),
        seq_lens=torch.tensor([n], dtype=torch.int32, device="cuda"),
        max_kv_len=n, is_causal=causal, return_lse=True, o_sf_scale=-1.0, out=out)


def flashinfer_ragged(q, k, v, causal=False):
    n, m = k.shape[0], q.shape[0]
    wr = flashinfer.prefill.BatchPrefillWithRaggedKVCacheWrapper(ws, "NHD")
    wr.plan(torch.tensor([0, m], dtype=torch.int32, device="cuda"),
            torch.tensor([0, n], dtype=torch.int32, device="cuda"),
            H, H, DQK, head_dim_vo=DV, causal=causal, sm_scale=SCALE,
            q_data_type=torch.bfloat16, kv_data_type=torch.bfloat16)
    return wr.run(q, k, v, return_lse=True)


# ---------------------------------------------------------------- part 1
qlen, kvlen = 8, 512
q = (torch.randn(qlen, H, DQK, device="cuda") * 0.5).bfloat16()
k = (torch.randn(kvlen, H, DQK, device="cuda") * 0.5).bfloat16()
v = torch.randn(kvlen, H, DV, device="cuda").bfloat16()
s = torch.einsum("qhd,khd->hqk", q.float(), k.float()) * SCALE
ref_lse_nat = torch.logsumexp(s, dim=-1).transpose(0, 1).contiguous()   # natural log

print("LSE base returned by each kernel (max |delta| against an fp32 reference):")
cands = [("trtllm_ragged_attention_deepseek", trtllm_ragged),
         ("flashinfer BatchPrefillWithRaggedKVCache", flashinfer_ragged)]
try:
    from sglang.jit_kernel.flash_attention_v4 import flash_attn_varlen_func as fa4

    def fa4_ragged(q, k, v, causal=False):
        return fa4(q=q, k=k, v=v,
                   cu_seqlens_q=torch.tensor([0, q.shape[0]], dtype=torch.int32, device="cuda"),
                   cu_seqlens_k=torch.tensor([0, k.shape[0]], dtype=torch.int32, device="cuda"),
                   max_seqlen_q=q.shape[0], max_seqlen_k=k.shape[0],
                   softmax_scale=SCALE, causal=causal, return_softmax_lse=True)

    cands.append(("fa4 (flash_attn.cute) varlen", fa4_ragged))
except Exception as e:  # optional
    print(f"  (fa4 not importable here: {type(e).__name__})")

for name, fn in cands:
    _, lse = fn(q, k, v)
    lse = lse.float()
    # kernels differ in layout: [tokens, heads] or [heads, tokens]
    if lse.shape == (H, qlen):
        lse = lse.transpose(0, 1).contiguous()
    d_nat = (lse - ref_lse_nat).abs().max().item()
    d_log2 = (lse - ref_lse_nat / LN2).abs().max().item()
    print(f"  {name:42s} vs natural log {d_nat:9.3g} | vs base-2 {d_log2:9.3g}"
          f"  -> base {'e' if d_nat < d_log2 else '2'}")

# ---------------------------------------------------------------- part 2
qlen, plen, slen = 8, 3072, 64        # long cached prefix, short new turn
q = (torch.randn(qlen, H, DQK, device="cuda") * 0.5).bfloat16()
k = (torch.randn(plen + slen, H, DQK, device="cuda") * 0.5).bfloat16()
v = torch.randn(plen + slen, H, DV, device="cuda").bfloat16()
s = torch.einsum("qhd,khd->hqk", q.float(), k.float()) * SCALE
ref = torch.einsum("hqk,khd->qhd", torch.softmax(s, dim=-1), v.float())

o_p, l_p = trtllm_ragged(q, k[:plen].contiguous(), v[:plen].contiguous())
o_s, l_s = trtllm_ragged(q, k[plen:].contiguous(), v[plen:].contiguous())
raw, _ = merge_state_v2(o_p, l_p, o_s, l_s)                  # what sglang does today
conv, _ = merge_state_v2(o_p, l_p * LN2, o_s, l_s * LN2)      # LSE converted to natural log

def rel(o):
    d = (o.float() - ref).abs()
    return d.max().item(), (d.mean() / ref.abs().mean()).item()

print("\nMerging a 3072-token prefix chunk with a 64-token suffix "
      "(merge_state_v2 assumes natural-log LSE):")
for name, o in (("raw base-2 LSE  (current behaviour)", raw),
                ("LSE * ln2       (converted)       ", conv)):
    m, r = rel(o)
    print(f"  {name}  max|delta| {m:.4g}  mean rel err {r:.3%}")

true_mass = torch.softmax(s, dim=-1)[:, :, plen:].sum(-1).mean().item()
mx = torch.maximum(l_p, l_s)
got = (torch.exp(l_s - mx) / (torch.exp(l_p - mx) + torch.exp(l_s - mx))).mean().item()
print(f"  softmax mass the suffix should get: {true_mass:.4%}; "
      f"weight the merge gives it: {got:.4%}")
```

Output on one B300 (sglang 0.5.15.post1, flashinfer-python 0.6.12, torch 2.11.0+cu130):

```
LSE base returned by each kernel (max |delta| against an fp32 reference):
  trtllm_ragged_attention_deepseek           vs natural log      2.79 | vs base-2  9.54e-07  -> base 2
  flashinfer BatchPrefillWithRaggedKVCache   vs natural log      2.79 | vs base-2  0.000244  -> base 2
  fa4 (flash_attn.cute) varlen               vs natural log  1.43e-06 | vs base-2      2.79  -> base e

Merging a 3072-token prefix chunk with a 64-token suffix (merge_state_v2 assumes natural-log LSE):
  raw base-2 LSE  (current behaviour)  max|delta| 0.008117  mean rel err 11.794%
  LSE * ln2       (converted)         max|delta| 0.0003404  mean rel err 0.267%
  softmax mass the suffix should get: 2.0326%; weight the merge gives it: 0.3719%
```

So on a long-prefix / short-suffix split — a chat turn appended to a cached conversation — the merged attention output is ~12% off, and the newest tokens receive about one fifth of the attention mass they should. The direction is consistent: the model under-attends the most recent tokens relative to the long cached prefix, and it repeats on every turn of a long conversation.

### Which backends reach the path

| backend | takes the chunk-and-merge path | LSE base | affected |
|---|---|---|---|
| `trtllm_mla` | every non-speculative extend with a cached prefix; no threshold, no one-shot variant | 2 | yes, on every cache hit |
| `flashinfer`, `flashmla` | only when the batch carries ≥ `chunked_prefix_cache_threshold` (8192) tokens of prefix **and** exceeds the one-shot chunk capacity | 2 | yes, above those sizes |
| `fa3` | same condition as `flashinfer` | e | no |
| `fa4` | every extend (`handle_attention_fa4` returns `MHA_CHUNKED_KV` unconditionally) | e | no |
| `triton` | never (absorbed MLA for any prefix) | n/a | no |

A production batch can put `flashinfer` on this path too: 32 requests of 4k context is already past the one-shot capacity. `fa3` returning natural-log LSE is presumably why this survived — `merge_state_v2` was ported from vLLM, where FlashAttention feeds it natural-log LSE, and `fa3` is the DeepSeek default on Hopper.

Decode is not affected: the absorbed-MLA decode path issues one paged kernel call and never returns an LSE to Python. Speculative target-verify is also off this path (`handle_attention_trtllm_mla` gates the merge on `is_extend_without_speculative()`).

### Suggested fix

Convert where the kernel returns the LSE, per backend, rather than inside the shared merge — a blanket conversion in `_chunked_prefix_attn_mha` would corrupt the already-correct `fa3` and `fa4` paths. That is also what vLLM does (`log2_lse_to_ln` in its own MLA prefill implementations). Concretely: scale by `ln 2` in the `trtllm_mla` ragged wrapper and in the flashinfer ragged prefill wrapper, or have those wrappers declare their LSE base and let `merge_state()` normalize on the declared base.

Happy to send a PR if the maintainers prefer one shape over the other.

### Environment

- sglang 0.5.15.post1 and 0.5.19 (runtime), `main` @ `2d25767` (code read)
- flashinfer-python 0.6.12 / 0.6.18, flash-attn-4 4.0.0b15, sgl-kernel from the matching wheels
- torch 2.11.0+cu130, CUDA 13.0, driver 580.126.20
- 8 × B300 SXM6, TP8, `DeepseekV3ForCausalLM`, bf16 weights, page_size 64, `chunked_prefill_size` 16384

**Comment by acmore** [comment]
Fix in #40927
